source/extensions/omni.isaac.lab/test/controllers/test_operation_space.py

```diff
@@ -0,0 +1,759 @@
+                ee_target_b, ee_target_pose_b, ee_target_pose_w, current_goal_idx = self._update_target(
+                    opc, root_pose_w, ee_pose_b, target_set, current_goal_idx
+                )
+                # set the opc command
+                opc.reset()
+                opc.set_command(ee_target_b, ee_pose_b)
+            else:
+                # get the updated states
+                jacobian, mass_matrix, gravity, ee_pose_b, ee_vel_b, root_pose_w, ee_pose_w, ee_force_w = (
+                    self._update_states(robot, ee_frame_idx, arm_joint_ids)
+                )
+                # compute the joint commands
+                joint_efforts = opc.compute(
+                    jacobian=jacobian,
+                    ee_pose=ee_pose_b,
+                    ee_vel=ee_vel_b,
+                    ee_force=ee_force_w,
+                    mass_matrix=mass_matrix,
+                    gravity=gravity,
+                )
+                robot.set_joint_effort_target(joint_efforts, joint_ids=arm_joint_ids)
+                robot.write_data_to_sim()
+
+            # update marker positions
+            self.ee_marker.visualize(ee_pose_w[:, 0:3], ee_pose_w[:, 3:7])
+            self.goal_marker.visualize(ee_target_pose_w[:, 0:3], ee_target_pose_w[:, 3:7])
+
+            # perform step
+            self.sim.step(render=False)
+            # update buffers
+            robot.update(sim_dt)
+
+    def _update_states(
+        self,
+        robot: Articulation,
+        ee_frame_idx: int,
+        arm_joint_ids: list[int],
+    ):
+        ee_jacobi_idx = ee_frame_idx - 1
+        # obtain quantities from simulation
+        jacobian = robot.root_physx_view.get_jacobians()[:, ee_jacobi_idx, :, arm_joint_ids]
+        mass_matrix = robot.root_physx_view.get_mass_matrices()[:, arm_joint_ids, :][:, :, arm_joint_ids]
+        gravity = robot.root_physx_view.get_generalized_gravity_forces()[:, arm_joint_ids]
+
+        # Compute current pose of the end-effector
+        ee_pose_w = robot.data.body_state_w[:, ee_frame_idx, 0:7]
+        root_pose_w = robot.data.root_state_w[:, 0:7]
+        ee_pos_b, ee_quat_b = subtract_frame_transforms(
+            root_pose_w[:, 0:3], root_pose_w[:, 3:7], ee_pose_w[:, 0:3], ee_pose_w[:, 3:7]
+        )
+        ee_pose_b = torch.cat([ee_pos_b, ee_quat_b], dim=-1)
+
+        # Compute the current velocity of the end-effector
+        ee_vel_w = robot.data.body_vel_w[:, ee_frame_idx, :]  # Extract end-effector velocity in the world frame
+        root_vel_w = robot.data.root_vel_w  # Extract root velocity in the world frame
+        relative_vel_w = ee_vel_w - root_vel_w  # Compute the relative velocity in the world frame
+        ee_lin_vel_b = quat_rotate_inverse(robot.data.root_quat_w, relative_vel_w[:, 0:3])  # From world to root frame
+        ee_ang_vel_b = quat_rotate_inverse(robot.data.root_quat_w, relative_vel_w[:, 3:6])
+        ee_vel_b = torch.cat([ee_lin_vel_b, ee_ang_vel_b], dim=-1)
```

Comment:
NOTE here that the linear velocity that you get is not necessarily at the frame of root/ee_frame. the output is at the COM of each of those RigidBodies.

Reply:
Since the COM of the ee is likely at or near the link frame this isn't really an issue. Just wanted you to be aware of it.

Reply:
It is a good point, and I believe it would be nice to correct this. Do you know how to do this? I do not know how to get COM frames.

Btw, I saw a similar problem with the Jacobian here #967, I guess the same problem applies for OPC. Hence, I took the solution there and adapted opc test code with the following:
```

# obtain dynamics related quantities from simulation
ee_jacobi_idx = ee_frame_idx - 1
jacobian_w = robot.root_physx_view.get_jacobians()[:, ee_jacobi_idx, :, arm_joint_ids]
mass_matrix = robot.root_physx_view.get_mass_matrices()[:, arm_joint_ids, :][:, :, arm_joint_ids]
gravity = robot.root_physx_view.get_generalized_gravity_forces()[:, arm_joint_ids]
# Convert the Jacobian from world to robot base frame
jacobian_b = jacobian_w.clone()
root_rot_matrix = matrix_from_quat(quat_inv(robot.data.root_state_w[:, 3:7]))
jacobian_b[:, :3, :] = torch.bmm(root_rot_matrix, jacobian_b[:, :3, :])
jacobian_b[:, 3:, :] = torch.bmm(root_rot_matrix, jacobian_b[:, 3:, :])
```

Reply:
Ah, I have just seen #966; we can wait for that to be merged for the velocities.